neet/sensitivity.py

```diff
@@ -35,33 +39,178 @@ def sensitivity(net, state):
+        if states is None:
+            states = list(net.state_space())
+        else:
+            states = list(states)
+
+        if weights is None:
+            weights = np.ones(len(states))
+        else:
+            weights = list(weights)
+
+        if np.shape(weights) != (len(states),):
+            raise(ValueError("Weights must be a 1D array with length same as states"))
+
+        norm = np.sum(weights)
+        for i,state in enumerate(states):
+            Q += weights[i] * difference_matrix(net, state, trans) / norm
+
+    else: # make use of sparse connectivity to be more efficient
+        state0 = np.zeros(N,dtype=int)
+
+        for i in range(N):
+            nodesInfluencingI = _connections(net,i)
+            for jindex,j in enumerate(nodesInfluencingI):
+            
+                # for each state of other nodes, does j matter?
+                otherNodes = list(copy.copy(nodesInfluencingI))
+                otherNodes.pop(jindex)
+                otherNodeStates = _states_limited(otherNodes,state0)
+                for state in otherNodeStates:
+                    # (might be able to do faster by calculating transitions once
+                    #  for each i)
+                    # (also we only need the update for node i)
+                    # start with two states, one with j on and one with j off
+                    jOff = copy.copy(state)
+                    jOff[j] = 0
+                    jOffNext = net.update(jOff)[i]
+                    jOn = copy.copy(state)
+                    jOn[j] = 1
+                    jOnNext = net.update(jOn)[i]
+                    # are the results different?
+                    Q[i,j] += (jOffNext + jOnNext)%2
+                Q[i,j] /= float(len(otherNodeStates))
+            
+    return Q
+
+def lambdaQ(net,**kwargs):
+    """
+    Calculate sensitivity eigenvalue, the largest eigenvalue of the sensitivity
+    matrix average_difference_matrix.
+    
+    This is analogous to the eigenvalue calculated in
+    
+        Pomerance, A, E Ott, M Girvan, and W Losert. ``The Effect of Network
+        Topology on the Stability of Discrete State Models of Genetic Control.''
+        Proc. Natl. Acad. Sci. USA 106, no. 20 (2009): 8209-14.
+        http://www.pnas.org/content/106/20/8209.full.
+    """
+    Q = average_difference_matrix(net,**kwargs)
+    return max(np.sort(abs(linalg.eigvals(Q))))
+
+def _fast_encode(state):
+    """
+    Quickly find encoding of a binary state.
+    
+    Same result as net.state_space().encode(state).
+    """
+    # see https://stackoverflow.com/questions/12461361/bits-list-to-integer-in-python
+    out = 0
+    for bit in state[::-1]:
+        out = (out << 1) | bit
+    return out
+
+def _boolean_distance(state1,state2):
     """
     Boolean distance between two states.
     """
-    return np.sum(abs(np.array(state1)-np.array(state2)))
+    out = 0
+    for i in range(len(state1)):
+        out += (state1[i]+state2[i])%2
```

Comment:
xor is Boolean addition, so you might change this to `out += state1[i] ^ state2[i]`.

I know that you like using `numpy`, so this entire function can be written as
```python
def _boolean_distance(state1, state2):
    return np.sum(np.logical_xor(state1, state2))`.
```
I can't speak for performance though.

Reply:
I now realize this is only used in calculating the sensitivity of an individual state, since we now use difference_matrix to calculate the average_sensitivity.  But who knows when we might use it in the future, so I ran some profiling.  The final score: xor wins at 2.5 s, mod addition 3.3 s, numpy loses at 4.2 s.

Reply:
My guess is that numpy loses by so much because `np.logical_xor` makes a copy.

There are a few other places where addition modulo 2 could be replaced with bitwise operation, e.g. lines 79, 91 and 164.